### DISCUSSION TITLE
I am a GAMMA software user, how should I organize the data so that I can use dolphin

### DISCUSSION THREAD

**Discussion by LDcrystal** [isce-framework/dolphin#260] [Q&A]
### Checked for duplicates

Yes - I've already checked

### Describe the bug

When I did [...] action, I noticed [...]

### What did you expect?

How should co-registered binary SLC be organized?

### Reproducible steps

```bash
1.
2.
3.
...
If you have a `dolphin_config.yaml` that you used, you can attach that as well.
```

### Environment

```bash
Please paste the output of the following:
`python -c "import dolphin; print(dolphin.show_versions())"`
...
```

**Comment by scottstanie** [comment]
I have not ever used Gamma (I don't have access to the software). 
Is there a description of what the outputs are after coregistering the SLCs?

I'll point out that currently `dolphin` can read in any GDAL-compatible rasters. So wherever your coregistered SLC stack is, you can point to it like 
```
dolphin config --slc-files ./path/to/*.slc.bin
```
Are you able to call `gdalinfo <name of slc file>.slc`?

A follow up- What satellite are you trying to process? If it's Sentinel-1, does Gamma organize them separately by bursts? By swath? Are they in radar or geo coordinates?

**Reply by songzwgithub** [reply]
Hello, I can provide the RSLC files generated by the GAMMA software in the radar coordinate system. Among them, the *.hgt file is the DEM in the radar coordinate system, and the *.lt_fine file is a refined lookup table (coordinate system A) containing pairs of real-valued output data coordinates from system B (fcomplex). If you have time to help test these files, please feel free to contact me if you have any questions. Thanks!
Here are the data links:
https://drive.google.com/file/d/1j0yXvr19wAepTnW5WNv-TQLac0axhCT8/view?usp=sharing

**Reply by scottstanie** [reply]
thanks! are those RSLCs already resampled so that the stack of images are coregistered? Or is this an intermediate step where you still need to use the lookup table to resample the SLCs?

**Reply by songzwgithub** [reply]
> thanks! are those RSLCs already resampled so that the stack of images are coregistered? Or is this an intermediate step where you still need to use the lookup table to resample the SLCs?

These RSLC images have been co-registered, without multi-looking. Within the GAMMA software, these RSLC products are now ready for subsequent interferogram generation. The lookup table file is prepared for geocoding.

**Reply by scottstanie** [reply]
I'm finding it difficult to find GAMMA documentation online- is this how you'd interpret the `SCOMPLEX` data format of theirs? as a flat binary file with real/complex values each as an `int16`?
```
data = np.fromfile('rslc_vv/20150110.vv.rslc', dtype=np.int16).reshape(600, 2000, 2)
slc = data[:,:,0] + 1j * data[:,:,1]
```

**Reply by songzwgithub** [reply]
> I'm finding it difficult to find GAMMA documentation online- is this how you'd interpret the `SCOMPLEX` data format of theirs? as a flat binary file with real/complex values each as an `int16`?
> 
> ```
> data = np.fromfile('rslc_vv/20150110.vv.rslc', dtype=np.int16).reshape(600, 2000, 2)
> slc = data[:,:,0] + 1j * data[:,:,1]
> ```

I haven't personally used Python to read GAMMA-generated SLC files, but I came across some sample code that might be useful. Hope this helps!
```python

  # read data
  cnt = ny*width
  if is_scomplex:
    scpx = np.fromfile(img, dtype=dt, count=2*cnt)
    scpx = np.reshape(scpx, (scpx.size//2, 2))
    data = np.empty((scpx.size//2,), dtype='c8')
    data = 1j*scpx[...,1];  data += scpx[...,0]
    data = data.astype('>c8')
    del(scpx)
  else:
    data = np.fromfile(img, dtype=dt, count=cnt)
  
```

**Reply by scottstanie** [reply]
I was able to run it on your data, though it looks the flat earth phase is still present? that's not something handled by Dolphin

<img width="570" alt="image" src="https://github.com/user-attachments/assets/33ab6e73-1bed-4bbb-a5b0-0df15ceb31ae" />
<img width="570" alt="image" src="https://github.com/user-attachments/assets/287bb4fa-c3a8-42fb-96f8-c632df0dc33c" />

**Comment by scottstanie** [comment]
You can try this script: https://github.com/scottstanie/dolphin/blob/2ecaf7cdf63254d4bec2da9c4accf522da549fe0/scripts/write-gamma-vrts.py 
I'll add it to dolphin at some point along with an FAQ or notebooks:

```bash
(mapping-312) staniewi:gamma-data$ python ~/repos/dolphin/scripts/write-gamma-vrts.py --par-files rslc_vv/*par
[vrt] rslc_vv/20150110.vv.rslc.vrt -> 20150110.vv.rslc (CInt16, 2000×600)
[vrt] rslc_vv/20150122.vv.rslc.vrt -> 20150122.vv.rslc (CInt16, 2000×600)
...
[vrt] rslc_vv/20160410.vv.rslc.vrt -> 20160410.vv.rslc (CInt16, 2000×600)
(mapping-312) staniewi:gamma-data$ gdalinfo rslc_vv/20150215.vv.rslc.vrt
Driver: VRT/Virtual Raster
Files: rslc_vv/20150215.vv.rslc.vrt
       rslc_vv/20150215.vv.rslc
Size is 2000, 600
Corner Coordinates:
Upper Left  (    0.0,    0.0)
Lower Left  (    0.0,  600.0)
Upper Right ( 2000.0,    0.0)
Lower Right ( 2000.0,  600.0)
Center      ( 1000.0,  300.0)
Band 1 Block=2000x1 Type=CInt16, ColorInterp=Undefined
```

then config and run like normal:
```
(mapping-312) staniewi:gamma-data$ dolphin config --slc-files rslc_vv/*.vrt --strides 5 2

(mapping-312) staniewi:gamma-data$ dolphin run dolphin_config.yaml
[INFO|wrapped_phase|L63] 2025-04-17T17:21:07-0400: Running wrapped phase estimation in /Users/staniewi/repos/dolphin/docs/notebooks/gamma-data
[WARNING|wrapped_phase|L490] 2025-04-17T17:21:07-0400: Could not make nodata mask: /Users/staniewi/repos/dolphin/docs/notebooks/gamma-data/rslc_vv/20160410.vv.rslc.vrt is not a CSLC file
[INFO|wrapped_phase|L107] 2025-04-17T17:21:07-0400: Creating persistent scatterer file /Users/staniewi/repos/dolphin/docs/notebooks/gamma-data/PS/ps_pixels.tif
```

**Reply by songzwgithub** [reply]
Thank you for providing the script. I encountered no issues when running it, but I have a few questions:

1. In GAMMA software, the "flat earth phase" removal is applied after generating the interferometric phase. Does this mean it is impossible to use Dolphin to process the Gamma-generated raslc files?
2. The output files lack geocoding. Is there a way to address this?
3. It seems there is no multi-looking step in the process. If we do not use full resolution, would applying multi-looking to the interferogram pairs reduce processing time?
4. Can MintPy be used for time-series processing with this workflow?

**Reply by scottstanie** [reply]
1. Not impossible (it ran though fine!),  but I'm assuming you'll want to remove it for analysis? I'm sure GAMMA has a way to remove that during coregistration.
2. I'm not sure how geocoding usually works for GAMMA. I know ASF HyP3 uses Gamma, and it looks like they [may have a module for it](https://github.com/ASFHyP3/hyp3-gamma/blob/develop/hyp3_gamma/insar/unwrapping_geocoding.py)
3. The phase linking does do multilooking during the covariance estimation, dolphin handles the output pixel spacing using the `strides` option: In my example, I did `dolphin config --strides 5 2 ...`, which means the `x` (column) size is 5x smaller, and the rows (azimuth) are 2x coarser. This makes it run ~10x faster, but it's up to you what output posting you want.
4. Yes - though I haven't used this script in awhile, there's [this "prep dolphin" script](https://github.com/OPERA-Cal-Val/calval-DISP/blob/00524195c7448ccc883d30cf9597dd2130c12a0d/dolphin_PSTworkflow/prep_mintpy_dolphin_timeseries.py#L531) in the opera science team repo

**Reply by songzwgithub** [reply]
Thank you again. I will proceed to try out the suggestions you provided.

**Reply by saboyce** [reply]
Were you able to establish a workflow to use dolphin with gamma files? I am interested in knowing whether the flat earth phase can be removed during coregistration and how the geocoding was handled.

**Reply by songzwgithub** [reply]
> Were you able to establish a workflow to use dolphin with gamma files? I am interested in knowing whether the flat earth phase can be removed during coregistration and how the geocoding was handled.

For flat earth phase removal, you can refer to the log section of the import file in this link:
[https://github.com/kanglcn/moraine/blob/main/nbs/Tutorials/CLI/load_data.ipynb](https://github.com/kanglcn/moraine/blob/main/nbs/Tutorials/CLI/load_data.ipynb)

**Reply by scottstanie** [reply]
Thanks @songzwgithub ! 
Just to repost: here it looks like he's running an offset calculation, then an phase simulator from each orbit using Gamma comands:
```python
    for i,(date,rslc,rslc_par) in enumerate(zip(dates,rslcs,rslc_pars)):
        off_par = scratch_dir/(reference+'_'+date+'.off')
        create_offset_command = f'create_offset {str(ref_rslc_par)} {str(rslc_par)} {str(off_par)} 1 1 1 0 >> {str(scratch_dir/"gamma.log")}'
        logger.info('run command: ' + create_offset_command)
        os.system(create_offset_command)
        sim_orb = scratch_dir/(reference+'_'+date+'.sim_orb')
        phase_sim_orb_command = f'phase_sim_orb {str(ref_rslc_par)} {str(rslc_par)} {str(off_par)} {str(hgt)} {str(sim_orb)} {str(ref_rslc_par)} - - 1 1 >> {str(scratch_dir/"gamma.log")}'
        os.system(phase_sim_orb_command)
        sim_orbs.append(sim_orb)
```
Then, the simulated phase is read and an cross multiplied by the SLCS `np.exp(1j * sim_orb)*rslc` . These should be usable in dolphin once you save the results.

**Comment by renierlgv** [comment]
Following up on this discussion with a similar problem. How to process other sensors data with dolphin? will it directly ingest coregistered slc from isce2 for example? If so, can you give a bit of a "walk trough" on how/what to do?

**Reply by scottstanie** [reply]
Yep dolphin should work! anything that is GDAL readable that's passed to the `--slc-files` will be accepted if it has some date(time) in the name. So far I have had success running it on Sentinel-1, NISAR, ALOS, and Capella SLCs. You'll just need to ensure the `wavelength` input option gets populated for unfamiliar preprocessors.


@mirzaees made this nice full walkthrough for running dolphin on isce2 results for the Earthscope ISCEPLUS short course

https://github.com/isceplus/2025-isceplus/blob/main/5.3_Intro_to_Dolphin/isce2-dolphin/dolphin-isce2.ipynb

The "deramping" shown above does need to happen before dolphin, as we haven't built in anything for coregistering+flattening SLCs due to the wealth of options out there

**Comment by renierlgv** [comment]
Thank you. My question aligns with @mirzaees case, ie, ISCE2 processed SLC from other sensors. However, I see no deramping step on  Sara's walkthrough, which make me believe that the Orlean data used there is already corrected as ISCE2 does the flattening at the interferogram level, right? meaning that I still have to do it myself using the ISCE2 generated range offset files or one of the options out there. Do you have any one in mind that you prefer/recommend?

**Reply by scottstanie** [reply]
Oh I see, perhaps it's a good idea to note in the docs that Dolphin accepts any GDAL-readable stack, but you need to coregister and phase-flatten them yourself.

Are you using one of the ISCE2 "contrib" workflows? I believe they all flatten phase during coregistration.

**Comment by renierlgv** [comment]
Noted. I just realized that the ISCE2 conda installation as from @lijun99 lacks the stack folder within contrib workflows you get when compiling isce2 from source (which somehow get stuck in the first step script when trying to create a slc stack for me), so, I'll carefully check the slc stack scripts and see whether or not they do deramp at slc level as it was my understanding it happens at interferogram level. thanks for the help!!

**Reply by lijun99** [reply]
I added instructions to install stack processors in isce2 a while ago. 

In ISCE2/topsStack, the flat earth phase can be removed either in generateIgram step or in resamp_withCarrier step (the default is latter in topsStack while the former in topsApp.py), search `flat`.

**Reply by renierlgv** [reply]
Thank you. I looked into the codes and it seems to be the case also for stripmap stacks (ie, resampleSlc.py) with flatten defaulting to True (flatten=(not inps.noflat)). can you confirm? for tops stacks I dont need to go trough ISCE2 as ISCE3/COMPASS handles it. I appreciate your time and help and accept my apologize if some questions are rather obvious as I'm not familiar (yet) with isce.

**Reply by lijun99** [reply]
That is correct.
